Function sections

gretay-js opened the pull request:
Add `--enable-function-sections` option to configure. With this option, the compiler will emit each function in a separate named text section, on supported targets. This enables function reordering using a linker script. With this option, the compiler emits additional labels `caml_hot__code_begin` and `caml_hot__code_end` in separate sections. This allows a linker script to move function sections outside of the segments they belong to, without breaking `caml_code_segments` table and the invariants it maintains. This way, functions from different compilation units can be reordered. 

The motivation for this patch is to improve layout of functions in large binaries. Hot functions that frequently call each other can be placed together in the hot segment in a way that reduces iTLB cache misses. See below for evaluation using Coq as a benchmark. 

All tests pass on precheck with function sections enabled (forced directly in configure for testing purposes).

## Supported targets

Support for --emit-function-section is implemented for amd64, i386, arm, arm74 targets on Linux systems.

MacOSX not supported: the "segment" part of section name is limited to 16 characters, which is too short for function names. We may be able to support it in the future if function names in generated assembly are replaced by identifiers.

Windows systems (32/64) not supported: NT.asm provide named text sections using .CODE directive. Cygwin and Mingwin allow named sections, but section arguments differ from standard Linux systems.

These are not conceptual limitations. Support on other targets/systems can be added later if desired. They were not included in this PR, because they require more testing of linker scripts and performance evaluation.

## Some design decisions

Making this a configure-time option has several advantages.

When this configure option is disabled, there is no change to generated assembly.  Tested on precheck with function sections disabled (the default) to confirm it. Precheck build # 228. Failure on ocam-freebsd-64 (git problem) is unrelated to this patch.

When function sections are enabled at configure time, functions in the runtime system and libraries are also emitted in separate function sections. These function are often amongst the hottest. Now they can be reordered alongside user-defined functions.

To reorder functions outside of their original segments, we emit an extra pair of labels `caml_hot__code_begin` and `caml_hot__code_end`. They are used at the beginning of `code_segment_table`. Each of these two labels is emitted in its own text section. It allows us to move these labels out from their original place using a linker script, and place other function sections between them. Any function section placed between them will be recognized as code from the code segment table.

The logic in asmlink that adds these labels to the code segments table is not target specific.  The hot sections and labels themselves are defined in `runtime/*.S` files for the supported targets. An alternative
would be to emit their definitions from `asmlink`. It would entail larger changes to `Emit` of all targets, even those in which function sections are not yet supported, because there is not facility currently to emit code outside of `Emit.begin_assembly` and `Emit.end_assembly` of a compilation unit. In any case, we had to make changes to `runtime/*.S` files to allow reordering of functions defined in these files.

To reorder C functions, C files in user code must be compiled with `-ffunction-sections`.  This compilation flag is added automatically at configure time and therefore C stubs in the ocaml runtime and libraries are compiled with it by default.

It is possible to link files some of which were compiled with function sections, and some without, but functions without sections cannot be moved.

Code size: Function sections increase size of intermiediate files. Using linker script as described below removes function sections in the final binary, therefore the size of the final binary is not affected by this PR (except the two new labels we add to make the hot code segment).

Profiling: This PR does not address the question of determining a good function order.  Below we describe a simple manual method of obtaining and applying profile information to choose function order. A proposal will be put forward separately for a better integration.  This PR is an initial step towards enabling hot/cold code reordering and related layout optimizations. The technical changes proposed here are enough to improve performance in critical systems.

## How to reorder functions using a linker script.

The following steps were tested on CentOS 7 with the GNU linker ld.

1. Obtain linker script by running the command `ld --verbose`.  Save output to file. Delete extra lines, usually everything before and after `===` lines, including these markers.

2. Patch linker script. The changes should look like this:
```
   .text           :
   {
     *(.text.unlikely .text.*_unlikely .text.unlikely.*)
     *(.text.exit .text.exit.*)
     *(.text.startup .text.startup.*)
     *(.text.hot .text.hot.*)
+   . = ALIGN(CONSTANT (MAXPAGESIZE)); 
+    *(.text.caml_hot__code_begin)
+    INCLUDE linker-script-hot
+    *(.text.caml_hot__code_end)
     *(.text .stub .text.* .gnu.linkonce.t.*)
     /* .gnu.warning sections are handled specially by elf32.em.  */
     *(.gnu.warning)
   }
```

3. Create linker-script-hot file that looks like this:
```
*(.text.camlTest__f3_44)
*(.text.camlTest__f1_28)
*(.text.camlTest__f0_19)
```
Thr file linker-script-hot must exists, but can be empty. Any function that is not listed here will remain in its original place.

6. Recompile the program with extra option `-ccopt -Wl,--script=linker-script` or rerun only the final link command with extra option `-Wl,--script=linker-script`. 
Check function order in a binary:
```
nm -n <test.exe> | grep ' T caml' | cut -c 20-
```
7. The order of functions in linker-script-hot can be chosen manually
from expert knowledge or generated automatically from perf profile or
using other tools.

For example:

Collect exection profile using `perf` :
```
perf record -e cycles:u -j any,u -o perf.data <test.exe> <args..>
```
Print execution counts:
```
perf report --stdio  -i perf.data --show-nr-samples
```
Create linker-script-hot based on execution counts:
```
perf report --stdio  -i perf.data --show-nr-samples -F sample,symbol -d <test.exe> | sed /^#/d | sed 's/^.*\[.\] //' | awk '!a[$0]++'  | grep -v @plt | head -n 1000 | sed 's/^/*(.text./' | sed 's/$/)/' > linker-script-hot
```

## Evaluation on `coq` as benchmark

1) "coqtop-nsatz" is one of the tests from coq test-suite.
2) "coqc-heap"  is one of the commands invoked by "compcert make".
3) "compcert-make" the entire build of compcert from scratch.

Perf data was obtained from "coqtop-nsatz" benchmark.

Results:
```
┌──────────────────┬───────────────────┬───────┬───────────────────────────┬───────┬─────────────┬───────┐
│                  │ coqtop-nsatz.orig │  +-   │ coqtop-nsatz.func-reorder │  +-   │  Difference │     % │
├──────────────────┼───────────────────┼───────┼───────────────────────────┼───────┼─────────────┼───────┤
│ iTLB-load-misses │         2,651,706 │ 2.43% │                 2,412,769 │ 3.01% │    -238,937 │ 0.910 │
│ L1-icache-misses │       619,155,761 │ 0.62% │               592,435,284 │ 0.08% │ -26,720,477 │ 0.957 │
│     cache-misses │       188,817,253 │ 0.52% │               188,575,119 │ 0.53% │    -242,134 │ 0.999 │
│           cycles │   147,074,415,845 │ 0.05% │           147,577,423,771 │ 0.40% │ 503,007,926 │ 1.003 │
│     instructions │   300,333,456,398 │ 0.01% │           300,434,632,303 │ 0.02% │ 101,175,905 │ 1.000 │
│ msec task-clock  │        35,369.081 │ 0.08% │                35,476.486 │ 0.39% │     107.405 │ 1.003 │
└──────────────────┴───────────────────┴───────┴───────────────────────────┴───────┴─────────────┴───────┘
```
```
┌──────────────────┬─────────────────┬───────┬─────────────────────────┬───────┬────────────────┬───────┐
│                  │  coqc-heap.orig │  +-   │  coqc-heap.func-reorder │  +-   │     Difference │     % │
├──────────────────┼─────────────────┼───────┼─────────────────────────┼───────┼────────────────┼───────┤
│ iTLB-load-misses │      19,013,641 │ 1.70% │              14,742,907 │ 9.61% │     -4,270,734 │ 0.775 │
│ L1-icache-misses │   4,617,532,849 │ 0.47% │           4,458,987,507 │ 0.31% │   -158,545,342 │ 0.966 │
│     cache-misses │     287,448,208 │ 0.24% │             289,574,916 │ 2.35% │      2,126,708 │ 1.007 │
│           cycles │ 164,311,594,187 │ 0.05% │         163,184,914,756 │ 0.42% │ -1,126,679,431 │ 0.993 │
│     instructions │ 253,059,095,057 │ 0.02% │         253,805,581,635 │ 0.33% │    746,486,578 │ 1.003 │
│       task-clock │      39,539.186 │ 0.03% │              39,282.762 │ 0.51% │       -256.424 │ 0.994 │
└──────────────────┴─────────────────┴───────┴─────────────────────────┴───────┴────────────────┴───────┘
```
Around 9% reduction in iTLB misses. The attached heatmaps visualize improvements in code locality.

Changes in execution times and related counters are in the noise, as can be seen from the correlation of % column and +- columns that show variance computed by perf -r for repeated runs. 

Changes in icache-misses may be just noise. It is a relatively small binary (18MB) running on a machine with a large 32MB icache. Large binary files we evaluated internally show consistent reduction in icache misses and execution time.

### Heatmaps

Each circle in a heatmap represents a block of addresses in the code segment of a binary.  The size and color of the circle shows how many times instructions in the corresponding block of addresses were accessed (as observed by perf with sampling).  Red circles are the most frequently accessed addresses. White dots are addresses that were never sampled.

These heatmaps show how function reordering packs together hot code to use about 1 MB of space, instead of the original range spanning the entire code section. There is still some activity outside the dense hot area, but they are relatively cold.

Original "coqtop-nsatz":
![coqtop-nsatz orig heatmap](https://user-images.githubusercontent.com/45005955/54612787-7a873b00-4a51-11e9-80f9-feb0cd5f21fa.png)

Optimized "coqtop-nsatz"::
![coqtop-nsatz func-reorder heatmap](https://user-images.githubusercontent.com/45005955/54612796-8115b280-4a51-11e9-9a7f-898c080c74cc.png)

Original coqtop from "compcert-make":
![coqtop-compcert-make orig heatmap](https://user-images.githubusercontent.com/45005955/54612811-870b9380-4a51-11e9-9fc3-4f1b100b6a0e.png)

Optimized coqtop from "compcert-make":
![coqtop-compcert-make func-reorder heatmap](https://user-images.githubusercontent.com/45005955/54612826-8f63ce80-4a51-11e9-9c64-8aad6da2c75b.png)

Optimized "coqc-heap":
![coqc-heap func-reorder heatmap](https://user-images.githubusercontent.com/45005955/54612838-9559af80-4a51-11e9-9c64-173afa8d84a8.png)

Heatmaps collected using FB BOLT tool.

Experiments on other large binaries also show execution time improvements.

gretay-js committed:
Emit each function in a separate section (amd64,i386,arm,arm64)

Add --enable-function-sections option to configure. With this option,
the compiler will emit each function in a separate named text section,
on supported targets. This enables function reordering using a linker
script. With this option, the compiler also emits caml_hot__code_begin
and caml_hot__code_end sections. This allows a linker script to
move function sections outside of the segments they belong to,
without breaking caml_code_segments.

gretay-js commented:
ping?

pmetzger commented:
@gretay-js A shame no one is picking up the ball and reviewing this. Maybe you should bring the work up on discourse or the mailing list?

gretay-js commented:
@pmetzger Thank you for the suggestion! The patch is shorter than the description :)

chambart commented on testsuite/tests/asmcomp/func_sections.ml:
```diff
@@ -0,0 +1,57 @@
+  ignore (test1 ());
+  ignore (test2 ());
+  ignore (test3 ());
+  ()
```
I don't see what that test is supposed to test.
I suppose that you have to inspect the generated assembly to see if it works as expected. If it is the case, maybe some script that messes with the result can be added. Like linking using some weird linker script.
Don't bother if this is too complex.

gretay-js replied:
Yes, this test is useless without some script. I checked the results manually by looking at the assembly output (produced with -S) to see that it has named text sections. It is awkward to do with ocamltest but possible. I've just push a commit that does it c2bcd9a2e82ba1e1ed5fb12cf95f31b6117312ca . As you can see, there is a bit of a hack in func_sections.run to check the generated assembly, and it requires a few new variables to be added in ocamltest. 

I also used a linker script that reorders the functions and then checked using "nm" that the order of function symbols is as expected. I have a script for it but it is tricky because it needs to check that the tools for binary inspection are available. I don't think we need it in the testsuite, as function reordering is not part of the compiler.

gretay-js replied:
I realized that I can simplify the conditions of this test case using the configure option. The updated version of the test is in commit 798c4f6769007002e5a302739fd204153d06d5bf.

chambart commented on configure:
```diff
@@ -2814,6 +2819,11 @@ VERSION=4.09.0+dev0-2019-01-18
+
+
+
+
```
Why include this whitespace change ?

chambart commented on runtime/arm.S:
```diff
@@ -99,11 +99,27 @@ alloc_limit     .req    r11
 
+        TEXT_SECTION(caml_call_gc)
         .align  2
         .globl  caml_call_gc
```
The duplication of the name in the macro and `.globl` seems dangerous. Maybe a single macro for declaring both would be better.

gretay-js replied:
The name appeared several times, unfortunately, even before this patch.  I haven't fixed it before as I wanted to keep the changes in this PR to minimum and only relevant to function sections.

I have just pushed an update that cleans it up a bit.  It replaces all the uses of names at the start of functions with a macro, similarly to amd64 backend.  There is another appearance at the end of functions, but I don't know how to fix it nicely in the same macro.

Running it on precheck now.

gasche commented:
I had a short discussion with @damiendoligez on how to get the PR to move forward: it looks like nice work but we need to find people interested in the work to review it and push it forward, and this is not so easy to find. Among OCaml users, relatively few are very knowledgeable about the backend aspects, and relatively few are working on the very-large binary programs that could benefit from this PR. Finding people at the intersection is hard!

A very tentative plan that I made up is:
- On a not-too-young laptop the cache size is typically much smaller than the 32MiB reported above (`sudo lshw -C memory` on my machine reports a 3MiB L3 cache), so more programs could benefit from the change (`coqtop`, `why3` etc. are at 18Mib-13Mib, so they may be affected by cache misses). 
- So maybe I could convince someone using these medium-sized programs to test the PR on their favorite program (or do it myself), following the reproduction steps described above. If they get some performance benefits, they can report it here, and give clear arguments for the fact that there are multiple groups that would benefit from the feature.

gretay-js committed:
Add macro FUNCTION in arm.S and arm64.S to avoid duplicate use of function name

gretay-js committed:
Improve test for function sections

mshinwell commented:
Instead of having another `configure` option, I think we could probably have a compile-time option, which we would always enable when building `stdlib` and `otherlibs`.  Those directories contain the most likely pieces of OCaml code (amongst all code built as part of the compiler's build) that people would want to rearrange into a particular order.  This will help reduce combinatorial explosion of the number of OPAM switches.  It should not make any difference to the binaries produced -- I think they should actually be identical, we can check.

avsm commented:
I'm really keen to try this in the context of:

- MirageOS (where we can do some pretty deep rearrangement of code layout all the way down to the bootlevel) 
- Sandmark (http://bench.ocamllabs.io) support could potentially automate the exploration of linkage options using perf that @gretay-js described.

However, it's so much easier to make this happen if it's a compile-time option that rather than a configure-time option, since we could just define a dune compilation context and recompile all the source code systematically.  Maintaining two compilers is quite heavyweight for this due to the other combinatorial explosions (e.g. flambda or not) that result.

gretay-js commented:
@avsm @mshinwell thank you, I'll change from configure to compile option.

gretay-js commented:
@gasche 
> A very tentative plan that I made up is: [...]

Thank you, it would be great! I will try the benchmarks you suggested and I'll also try to find more external benchmarks.  
For running on a laptop for more constraint resources, note that the current PR doesn't support function sections on macos, where names of section segments are restricted to 16 characters, because function names are often longer, especially for anonymous functions.

gretay-js committed:
Add compile-time option -function-sections

gretay-js commented:
I added the new compile-time option -function-sections. 

Target support for function sections is detected automatically by configure. If function sections are not supported on the target, configure prints a message and the compiler rejects -function-sections as "not available. The user can disable function sections at configure time with --disable-function-sections. @avsm Is it a convenient default for avoiding too many configurations?

With function sections enabled at configure time, all executables will contain two unused symbols, caml_hot__code_begin and caml_hot__code_end (defined in libasmrun), even if they were compiled without -function-sections. This can be avoided by emitted the two symbols from asmlink, conditionally on -function-section compile option, instead of including them in the runtime assembly.  However, to implement it, we would need to change cmm and emitters.

When function sections are enabled at configure time, the compiler itself is built with function sections, including the standard library, runtime, and compiler libs.  

Function sections can increase the size of intermediate files, but the final executable size change is negligible. For example, the size of ocamlopt.opt increased by 52 bytes in my local build (amd64,Centos,with flambda), because of the two new symbols and the extra entry in the code segment table. The size of intermediate files (objects,lib) increased by just under 5%.  

When the compiler gets -function-sections compile-time option, it also adds -ffunction-sections to the flags passed to the C compiler, so that C code gets compiled with function sections as well.

gretay-js committed:
Fix arm emitters

gretay-js committed:
Fix test for arm targets

gretay-js commented:
@gasche  Here are the results for another benchmark: frama-c on idct example.
It shows a reduction of more than 10% in iTLB misses.
There is also a small reduction in icache misses, and a small improvement in execution time.
The executable size is 37MB. Benchmarked on a machine with 32KB L1 icache and 4K pages.
This much improvement in iTLB should translate into an improvement in execution time on more resource-constraint machines or with larger binaries.
```
┌──────────────────┬────────────────┬───────┬────────────────┬───────┬──────────────┬─────────┐
│                  │           orig │  +-   │   func-reorder │  +-   │   Difference │       % │
├──────────────────┼────────────────┼───────┼────────────────┼───────┼──────────────┼─────────┤
│ L1-icache-misses │    441,156,434 │ 0.30% │    423,769,824 │ 0.22% │  -17,386,610 │  -3.941 │
│           cycles │ 14,801,935,826 │ 0.19% │ 14,482,161,463 │ 0.12% │ -319,774,363 │  -2.160 │
│ iTLB-load-misses │      2,901,373 │ 5.65% │      2,587,688 │ 3.77% │     -313,685 │ -10.812 │
│       task-clock │      3,550.054 │ 0.20% │      3,473.881 │ 0.13% │      -76.173 │  -2.146 │
└──────────────────┴────────────────┴───────┴────────────────┴───────┴──────────────┴─────────┘

```
@shindere  Could you please review the changes in configure.ac?

gretay-js commented:
All tests pass on precheck (build # 284).

chambart commented:
LGTM. I won't comment on the autoconf part for which I'm not competent.

I'm quite convinced that this can have large run-time effects. This have been quite well documented in various articles. I won't ask for benchmarks. Moreover, there is basically no cost to this change.
It would be nice to provide a tool (a small shell script or a nice ocaml version) that profiles and produces the function order, and add some documentation to the manual. Otherwise this will probably go unnoticed.

It might be interesting to change a bit the name of the sections to something like `.text.caml.function_name`. This might be useful for other tricks and would allow to avoid needing the `caml_hot__code_begin` symbol.
Your linker script could instead be something like:
```
   . = ALIGN(CONSTANT (MAXPAGESIZE)); 
    *(.text.caml__code_begin)
    INCLUDE linker-script-hot
    *(.text.caml.*)
    *(.text.caml__code_end)
```

I don't really have anything against this symbol, It just feels weird that a symbol named caml_hot__code_begin emission is controled by a '--function-section" argument.

But seriously, JS, you are really producing some horrendously big binaries for that to matter...

gretay-js commented:
> It would be nice to provide a tool (a small shell script or a nice ocaml version) that profiles and produces the function order, 

I am working on just that.

> and add some documentation to the manual. Otherwise this will probably go unnoticed.
thanks, will do! 



> It might be interesting to change a bit the name of the sections to something like `.text.caml.function_name`. This might be useful for other tricks and would allow to avoid needing the `caml_hot__code_begin` symbol.

Changing the name is easy, I can do it, but how does the name change avoid needing the symbol?


> I don't really have anything against this symbol, It just feels weird that a symbol named caml_hot__code_begin emission is controled by a '--function-section" argument.

I agree. Would always having the symbol always around be better? 

> But seriously, JS, you are really producing some horrendously big binaries for that to matter...

all we need is whole program dead code elimination :)

shindere commented:
Greta Yorsh (2019/07/03 05:29 -0700):
> @shindere  Could you please review the changes in configure.ac?

Of course! Thanks for asking!

One high-level question I have is: why do we need configure-time support
for that? Why isn't a compiler-option enough?

An implementation-related comment is that if the feature is explicitly
requested but is actually not supported (because of the architecture,
compiler or whatever) then in my opinion this should be an error and not
a warning as is currently implemented. Others may have different
opinions here but I think so far that's the policy we had that if a
feature can't be supported but has been explicitly requested then
configure should just fail.

shindere commented:
Greta Yorsh (2019/07/03 05:30 -0700):
> All tests pass on precheck (build # 284).

Well did you enable the feature for the tests?

dbuenzli commented:
> One high-level question I have is: why do we need configure-time support for that? Why isn't a compiler-option enough?

What about no option at all ?

shindere commented:
Daniel Bünzli (2019/07/04 02:34 -0700):
> > One high-level question I have is: why do we need configure-time support for that? Why isn't a compiler-option enough?
> 
> What about no option at all ?

Well if this behaviour is considered strictly better than the former one
and it is fine to provide it when possilbe, if there is no reason one
may want to not have it then yes of course that could become the
default and no option would be needed neither at configure not at
compile time.

dbuenzli commented:
@shindere if that is the case then a configure time option feels better. It makes it easier to test the feature until it's been made the default (with the cli you need to tweak build systems).

shindere commented:
Daniel Bünzli (2019/07/04 02:46 -0700):
> @shindere if that is the case then a configure time option feels
> better. It makes it easier to test the feature until it's been made
> the default (with the cli you need to tweak build systems).

Makes sense. But then it'd mean that the `-function-sections` currently
implemented in this PR should be removed, right?

mshinwell commented:
To recap how this works:
1. `configure` tests whether the platform and C compiler support function sections.  (In particular they are not supported on macOS targets.)
2. The compiler libraries and runtime are built with function sections enabled, if the `configure` test succeeds.  The overhead in object file size will be absolutely minimal in this case and the compiler executables should in all material respects be identical---but crucially this gives the user the opportunity to specify the layout of pieces of code such as those in the runtime.
3. The flag is provided to the compiler as well so that users don't have to enable this on their own code bases unless they want to use it.  The object file size penalty may be larger here and undesirable except when performing optimized builds using function sections.  This flag is provided as a compile-time option rather than a configure-time option to avoid a proliferation of compiler OPAM switches, etc.

The main reason the `configure` flag exists is to disable the feature in the event that someone finds building the runtime and compiler with it enabled causes a problem, or similar.  This is unlikely (we think very unlikely) but it's worth having the escape hatch.

I agree with @shindere that failure of `configure` to detect a compiler and platform supporting function sections should be an error if the user explicitly requested that support.

@gretay-js We should adjust the patch so that there is `OCAMLPARAM` support for enabling function sections.

shindere commented:
Mark Shinwell (2019/07/04 04:49 -0700):
> The main reason the `configure` flag exists is to disable the feature
> in the event that someone finds building the runtime and compiler with
> it enabled causes a problem, or similar.

Well at the moment the feature is disabled by default, right?

gretay-js commented:
> Well at the moment the feature is disabled by default, right?

That was the case in the first version of the PR.  
In the current version, function sections are enabled by default at configure time (if the target supports it), but the compiler does not generate function sections by default.  Users need to invoke the compiler explicitly with "ocamlopt ... -function-sections" to get it. 

It is not quite the way typical configure/compiler options interact, but there are good reasons to have it this way, as @mshinwell summarized above.

gretay-js commented:
> Well did you enable the feature for the tests?

yes :)
There is also a new test that uses it, passing -function-sections flag to the compiler.  It even failed at first (on arm and macos that were misconfigured) and I fixed. It's great to have a test!

gretay-js commented:
>  if the feature is explicitly requested but is actually not supported (because of the architecture,
compiler or whatever) then in my opinion this should be an error and not a warning as is currently implemented.

I agree but couldn't figure out how to implement it.  Is there a way to check in configure.ac whether --enable-function-section was explicitly passed to configure or not?

gretay-js commented:
> It might be interesting to change a bit the name of the sections to something like `.text.caml.function_name`. This might be useful for other tricks and would allow to avoid needing the `caml_hot__code_begin` symbol.
> Your linker script could instead be something like:
> 
> ```
>    . = ALIGN(CONSTANT (MAXPAGESIZE)); 
>     *(.text.caml__code_begin)
>     INCLUDE linker-script-hot
>     *(.text.caml.*)
>     *(.text.caml__code_end)
> ```

Function names already start with "caml" so we can already filter them out similarly to the way you showed here. There aren't caml__code_begin/end symbols though for the whole code yet. The ones we have are per module and the functions get reordered outside of their original modules, so we don't know the end.

Are you suggesting to replace caml_hot__code_begin/end with caml__code_begin/end that are always emitted surrounding the entire code?  I like it. It looks like these symbols can be used (in a separate patch of course) to simplify code_segments_table as well.

gretay-js committed:
Add function-sections to OCAMLPARAM

shindere commented:
Greta Yorsh (2019/07/08 04:41 -0700):
> I agree but couldn't figure out how to implement it.  Is there a way
> to check in configure.ac whether --enable-function-section was
> explicitly passed to configure or not?

Yes. Consider for instance

```
AC_ARG_ENABLE([foo],...)
```

For such an option, there are three (well actually four if you use = but
let's leave this aside) cases:

 1. `./configure --enable-foo`: `enable_foo` is set to `yes`.

 2. `./configure --disable-foo`: `enable_foo` is set to `no`.

 3. `./configure`: `enable_foo` is by default set to the empty string,
 but you can customize this value. See for example how `AC_ARG_ENABLE`
 is called wiht the `debugger` argument. In this case, when nothing is
 specified on configure's command-line, `enable_debugger` is set to
 `auto`.

gretay-js committed:
Error in configure --enable-function-sections if target doesn't support it

If the user explicitly requests function sections by calling configure
with --enable-function-sections on a target that does not support it,
it's an error at configure time.

gretay-js committed:
Add .caml to function section names

Emit .text.caml.function_name instead of .text.function_name,
and update runtime assembly function names accordingly.

gretay-js committed:
No support for section names on macos/mingw/cygwin

Cleanup text section definition in runtime, now that function sections
are disabled from configure on these targets.

gretay-js commented:
a13d91b: If the user explicitly requests function sections by calling configure with --enable-function-sections on a target that does not support it, it's an error at configure time Thanks to @shindere for the explanation on how it's done.

11d99dc: Adds function-section to OCAMLPARAM, as @mshinwell suggested earlier.

c51c39b: Adds .caml to sections name prefix. Running it on precheck again. 

@chambart  To place caml__code_begin/end correctly around the whole code without a linker script, we would need quite a bit of change in asmlink, I think, especially if it is unconditional, not only for function sections. Is it okay to keep caml_hot__code_begin/end names for now? It's not too bad to have an empty "hot" code segment enabled by default at configure. It's not a compile-time option "function-section" that creates these symbols.

gretay-js committed:
Update expected output of func_sections test

gretay-js commented:
Precheck (build 286) results: there is one failure on cygwin-64 when flambda=false. It looks like a system crash unrelated to ocaml. Everything else is fine.

shindere commented:
Yeah these kind of crashes occur sometimes and it is very okay not to
take them too seriously.

dra27 commented:
Indeed - cygwin64 is worth paying attention to only because it's the only non-shared build in CI, but the fact it's passing on flambda=true suggests all is probably well.

gretay-js committed:
Changes entry

gretay-js commented:
I've just rebased and added Changes entry. If there are no other comments, can this be merged please?

chambart commented:
@gretay-js 
> @chambart To place caml__code_begin/end correctly around the whole code without a linker script, we would need quite a bit of change in asmlink, I think, especially if it is unconditional, not only for function sections. Is it okay to keep caml_hot__code_begin/end names for now? It's not too bad to have an empty "hot" code segment enabled by default at configure. It's not a compile-time option "function-section" that creates these symbols.

Yes I'm perfectly ok with it. The alternatives are probably too complex for what it brings.
It is good to go for me.

gasche reviewed (approved):
I am approving on behalf of Pierre Chambart's review (and Sébastien for the autoconf aspects) -- Pierre, please do not hesitate to approve yourself next time. I didn't have the time to do the measurements that I said I could do (it's actually a fair amount of low-level twiddling for a non-expert such as myself), but I understand the role and usefulness of the PR, and I like the fact that the PR is fairly small and non-invasive. I have decided to merge it.

One aspect of the PR that is a bit lacking is tooling support and/or a good workflow to actually make use of it. I appreciated the effort of @gretay-js to provide clear explanations (I could follow them, knowing nothing at all of linking intricacies), but the workflow provided is still very manual. So it is legitimate to wonder who in practice will make use of the feature (besides JS who expressed clear interest); in the current state of affairs, I don't see the Why3/Coq people having the knowledge in-house to use it in their build&packaging scripts. (I think that this perception, of a feature that only a few specialized shops could ever use, also made the review&decision longer than it could have been.)

It would be nice if we eventually had support for doing the layout choices in a more automated way, but I wonder if there already exist tools to do this that we could reuse; what about the existing open-source projects that do layout optimization (Firefox, Chrome, etc.), do they use a standardized toolchain?

gretay-js commented:
@gasche Thank you for approving and merging the PR! 
I am working on a tool that automates and extends the basic workflow described in this PR. The tool takes as input a profile produced by perf and automatically generates a linker script with hot functions layout. The tool also uses a profile to inform optimization decision made at compile time. Currently, only reordering of basic blocks within a function is implemented. The basic approach is similar to profile guided optimizations, but the profile is collected using sampling instead of instrumentation. There are tools like this for C/C++ (e.g. Google's [AutoFDO](https://github.com/google/autofdo) and Facebook's [BOLT](https://github.com/facebookincubator/BOLT)), but not for Ocaml.

mshinwell commented:
We'll give full details of the new tool in due course, but one thing I'd like to mention now is that we think it will have substantially higher correctness guarantees than existing tools in the same area.

copy commented:
Could function sections be used with the `--gc-sections` linker flag to get dead code elimination? I did a quick test with 4.10.0+beta1 and `-ccopt -Wl,--no-export-dynamic,--gc-sections,-print-gc-sections`:  C functions (for example from libasmrun.a) are removed, but OCaml functions aren't. I suspect OCaml functions are referenced in the data section, so something like -fdata-sections might be necessary, but my knowledge of compiler internals end here.

gretay-js commented:
That's right, there are references into ocaml functions from the data section, so they are not dead as far as the linker can tell.  Adding separate data sections per function would not be enough, because there are references from the frametable and the module block.
